pkg/broker/handler/pool/retry/pool.go

```diff
@@ -0,0 +1,150 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package retry
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"go.uber.org/zap"
+
+	"knative.dev/pkg/logging"
+
+	"github.com/cloudevents/sdk-go/v2/protocol/pubsub"
+	"github.com/google/knative-gcp/pkg/broker/config"
+	"github.com/google/knative-gcp/pkg/broker/handler"
+	handlerctx "github.com/google/knative-gcp/pkg/broker/handler/context"
+	"github.com/google/knative-gcp/pkg/broker/handler/pool"
+	"github.com/google/knative-gcp/pkg/broker/handler/processors"
+	"github.com/google/knative-gcp/pkg/broker/handler/processors/deliver"
+)
+
+// TODO Retry and Fanout are using similar SyncPool struct, needing an interface to reduce redundancies.
+// SyncPool is the sync pool for retry handlers.
+// For each trigger in the config, it will attempt to create a handler.
+// It will also stop/delete the handler if the corresponding trigger is deleted
+// in the config.
+type SyncPool struct {
+	options *pool.Options
+	targets config.ReadonlyTargets
+	pool    sync.Map
+}
+
+// StartSyncPool starts the sync pool.
+func StartSyncPool(ctx context.Context, targets config.ReadonlyTargets, opts ...pool.Option) (*SyncPool, error) {
+	p := &SyncPool{
+		targets: targets,
+		options: pool.NewOptions(opts...),
+	}
+	if err := p.syncOnce(ctx); err != nil {
+		return nil, err
+	}
+	if p.options.SyncSignal != nil {
+		go p.watch(ctx)
+	}
+	return p, nil
+}
+
+func (p *SyncPool) watch(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-p.options.SyncSignal:
+			if err := p.syncOnce(ctx); err != nil {
+				logging.FromContext(ctx).Error("failed to sync handlers pool on watch signal", zap.Error(err))
+			}
+		}
+	}
+}
+
+func (p *SyncPool) syncOnce(ctx context.Context) error {
+	var errs int
+
+	p.pool.Range(func(key, value interface{}) bool {
+		// Each target represents a trigger.
+		if _, ok := p.targets.GetTargetByKey(key.(string)); !ok {
+			value.(*handler.Handler).Stop()
+			p.pool.Delete(key)
+		}
+		return true
+	})
+
+	p.targets.RangeAllTargets(func(t *config.Target) bool {
+		tk := config.TriggerKey(t.Namespace, t.Broker, t.Name)
+		bk := config.BrokerKey(t.Namespace, t.Broker)
+
+		// There is already a handler for the trigger, skip.
+		if _, ok := p.pool.Load(tk); ok {
+			return true
+		}
+
+		opts := []pubsub.Option{
+			pubsub.WithTopicID(t.RetryQueue.Topic),
+			pubsub.WithSubscriptionID(t.RetryQueue.Subscription),
+			pubsub.WithReceiveSettings(&p.options.PubsubReceiveSettings),
+		}
+
+		if p.options.ProjectID != "" {
+			opts = append(opts, pubsub.WithProjectID(p.options.ProjectID))
+		} else {
+			opts = append(opts, pubsub.WithProjectIDFromDefaultEnv())
+		}
+
+		if p.options.PubsubClient != nil {
+			opts = append(opts, pubsub.WithClient(p.options.PubsubClient))
+		}
+		ps, err := pubsub.New(ctx, opts...)
```

Comment:
I forgot to mention. For retry, maybe you want to override the default pubsub `ReceiveSettings` - 1) you probably want to set `Synchronous` to `true` 2) you probably want to limit the `NumGoroutines`, `MaxOutstandingMessages` and `MaxOutstandingBytes`

Reply:
When `Synchronous` is true, then no more than `MaxOutstandingMessages` will be in memory at one time, `MaxOutstandingBytes` still refers to the total bytes processed, rather than in memory and `NumGoroutines` is ignored. Do you think I need to set the Synchronous as ENV in the main to check whether it is set to true or false, or directly override it to true and set `MaxOutstandingMessages` and `MaxOutstandingBytes` when building `buildPoolOptions`?

Reply:
Currently, I only set the `Synchronous` to `true`, but didn't change `MaxOutstandingMessages` and `MaxOutstandingBytes`

Reply:
Please add a TODO and we can revisit this topic.